navapbc/platform-test
# Add pull request environments

## Ticket

Resolves https://github.com/navapbc/template-infra/issues/139

## Changes

Add callable reusable github action workflows that:
- creates or updates a temporary environment in a separate workspace
- destroys the temporary environment and workspace

Add template github action workflows for an app that calls the above reusable workflows to:
- create or update a temporary dev environment when a PR is opened, updated, or reopened
- destroys the temporary environment when the PR is closed

Add documentation around PR environments

## Context for reviewers

This is a feature that has proven to be quite useful on a number of projects, enabling product managers and designers to see and collaborate with engineers on in progress features to UAT the feature or otherwise give feedback on it.

This feature will also enable things like end-to-end testing on PRs and accessibility checks on PRs by providing a full environment they can access that is agnostic to the underlying web stack.

## Testing

Note: Ignore the anchore scan failures since those aren't relevant to this PR

1. In creating this PR, the workflow should have created a PR environment and posted a comment on the PR with a link to the environment.

   Workflow run that created the workspace and PR environment: https://github.com/navapbc/platform-test/actions/runs/9667773253?pr=114

   See that a workspace was created for this PR
   <img width="621" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/2d829489-6759-4098-9642-d5708cba2a72">

   Screenshot of PR environment:
   <img width="636" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/007e668a-fc38-4176-bbe9-942cd1d4f711">

2. Make a second change to the hello world app to change the content, and see that it updates in the PR environment but not in the original dev environment at https://platform-test-dev.navateam.com/. This change is purely for testing and won't be included in the template-infra PR.

   Workflow run that created the PR environment: https://github.com/navapbc/platform-test/actions/runs/9667905228/job/26670778105

   Screenshot of PR environment:
   <img width="635" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/13a1c194-a145-42e0-9b5f-edf7d740cdf6">

3. Close the PR. The workflow posts a comment that the environment is destroyed, and 

   Workflow run that destroyed the environment and workspace: https://github.com/navapbc/platform-test/actions/runs/9668062116

   See that the workspace no longer exists
   <img width="633" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/862f3d17-bac3-48ca-840e-d0c40e254973">

4. Reopen the PR, and see that the environment is recreated at http://p-114-app-dev-1220465442.us-east-1.elb.amazonaws.com/ (note that the URL changes after destroying and recreating)
